Searchwork. Fixes #795. Also fixes #834

hallvors opened the pull request:
Here's a significant refactoring of the search/filter code. Search parameters are split out into a separate model. It should now be possible to combine label search, statuses at the top and free-text keywords.

Not quite ready to merge - I think there may still be a few bugs to find.

r?@miketaylr for the JS and @magsout for the styling.

hallvors commented:
Some Travis log "expecting fixtures file ..." makes no sense to me. The changes on this branch are already on top of latest master as far as I can tell, and none of these changes were about removing some fixture file(s)..

magsout commented on webcompat/templates/issue-list/applied-labels.html:
```diff
@@ -0,0 +1,11 @@
+  <div class="wc-content">
+    <script type="text/template" id="issuelist-appliedlabels-tmpl">
+      <% _.each(labels, function(label){ %>
+          <button type="button" class="wc-Filter wc-Filter-appliedlabel js-filter-button wc-label-<%= label %>">
```
In this case it's a modifier, so `wc-Filter--appliedLabel`

magsout commented on webcompat/static/css/development/components/applied-labels.css:
```diff
@@ -0,0 +1,11 @@
+  background-color:#F7F7F7;
+  padding-bottom:1.5em;
+}
+.wc-Filter-appliedlabel{
```
it's a modifier of `Filter`, so please add this properties in filter.css

hallvors commented:
These changes quite possibly need test updates. Perhaps the missing fixture file messages are due to API request URLs changed because of the new model reordered the query string or something?

hallvors commented:
I'm investigating the failures.. TBC

miketaylr commented:
> Some Travis log "expecting fixtures file ..." makes no sense to me.

That is logged every time, it doesn't mean an error (so I guess we should update the log) -- but it could be that the fixture files are out of sync now. I can take a look a bit later at the failures (and the rest).

miketaylr commented on webcompat/api/endpoints.py:
```diff
@@ -187,14 +187,21 @@ def get_category_from_search(issue_category):
+    query_string = ''
+
+    # restrict results to our repo.
+    query_string += " repo:{0} ".format(REPO_PATH)
```
nit: Can we use single quotes for consistency here?

miketaylr replied:
Actually, do we need this? All the possible return values of this method call `get_search_results` which does the exact same thing, so are we sending a double `repo:foo`?

miketaylr commented on webcompat/static/css/development/components/filter.css:
```diff
@@ -61,3 +61,10 @@
   }
+  .wc-Filter--appliedlabel{
+    padding-right: 40px;
+    background-image: url(/img/invalid.svg);
```
Ah, I see. You're just reusing the red X from the form validation (re: my comment in the bug about the red X). I wonder if we could find a nice icon from the same icon font set that we use elsewhere in the site. cc @magsout

magsout replied:
How I can see this class applied to the DOM ?

hallvors replied:
It's on the /issues page, there will be a new DIV wc-AppliedLabels just after wc-IssueSearch. If you click a label in the search results list (to the right of each issue), the DIV will have children. See screenshot in #795.

miketaylr commented on webcompat/static/js/lib/issue-list.js:
```diff
@@ -84,12 +84,14 @@ issueList.DropdownView = Backbone.View.extend({
   _isLoggedIn: $('body').data('username'),
   _userName: $('body').data('username'),
-  initialize: function() {
+  initialize: function(theOptions) {
```
nit: can we use `options` like you did up in `issueList.DropdownView#initialize`?

hallvors replied:
Your code already used options as a local variable name in this method so I wasn't allowed to :-p
Can change that of course..

miketaylr commented on webcompat/static/js/lib/issue-list.js:
```diff
@@ -191,42 +188,53 @@ issueList.SearchView = Backbone.View.extend({
-    }
+    this.mainView.params.setParam('search', searchValue);
+  }, 250),
+});
```
I guess we don't prevent doing search for the same thing, or an empty search now. Is that intentional? (maybe you do later on, I haven't read everything yet).

hallvors replied:
See line 37 onwards in query-params.js

miketaylr commented on webcompat/static/js/lib/issue-list.js:
```diff
@@ -191,42 +188,53 @@ issueList.SearchView = Backbone.View.extend({
+/*
+* The applied labels view shows the labels we're filtering by - *excluding*
+* status- prefixed labels. These labels can be un-set by clicking their X
+* or by focusing them by keyboard and pressing del
```
:100: for keyboard functionality.

miketaylr commented on webcompat/static/js/lib/issue-list.js:
```diff
@@ -191,42 +188,53 @@ issueList.SearchView = Backbone.View.extend({
+  initialize: function(options) {
+    this.mainView = options.mainView;
+    issueList.events.on('appliedlabels:update', _.bind(this.updateAppliedLabels, this));
+
```
nit: empty line

miketaylr commented on webcompat/static/js/lib/issue-list.js:
```diff
@@ -191,42 +188,53 @@ issueList.SearchView = Backbone.View.extend({
+    return this;
+  },
+  deleteLabel: function(e) {
+    var label = e.target.firstChild.textContent.trim();
```
Looks like `String#trim` works everywhere (IE9+). Another option would be `$.trim($(e.target).first().text())`, but I think I prefer your way (and I don't think we really "support" IE8 or below-- we certainly don't test on it).

hallvors replied:
If we're going to support old browsers I'd prefer adding String.prototype.trim=.. somewhere :)

miketaylr commented on webcompat/static/js/lib/issue-list.js:
```diff
@@ -299,56 +307,20 @@ issueList.IssueView = Backbone.View.extend(
-          if (urlParams !== $.param(this.issues.params)) {
-            this.updateModelParams(urlParams);
-          }
+      if (this._urlParams.length !== 0) {
```
:clock430:   (just leaving a bookmark for myself, I'll pick up review here tomorrow) :clock8:

miketaylr commented:
I ran tests locally and pretty much all the search and issue-list tests fail... 

```
FAIL: main - Reporting (non-auth) - URL validation (731ms)
FAIL: main - Reporting (non-auth) - Problem type validation (621ms)
(we can ignore these, they always fail on my Mac, but pass on Travis :/)


FAIL: main - Issue-list - IssueListView renders (11747ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".js-issue-list .wc-IssueItem"}] Unable to locate element: {"method":"css selector","selector":".js-issue-list .wc-IssueItem"}
FAIL: main - Issue-list - PaginationControlsView tests (10764ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".js-pagination-previous.is-disabled"}] Unable to locate element: {"method":"css selector","selector":".js-pagination-previous.is-disabled"}
FAIL: main - Issue-list - Pagination dropdown tests (11818ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-child(51)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-child(51)"}
FAIL: main - Issue-list - Loading issues page has default params in URL (10623ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1)"}
FAIL: main - Issue-list - Loading partial params results in merge with defaults (10671ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1)"}
FAIL: main - Issue-list - Dropdowns reflect state from URL (215ms)
AssertionError: Pagination dropdown label is updated from URL params: expected 'Show 50' to equal 'Show 25'
FAIL: main - Issue-list - Going back in history updates issue list and URL state (203ms)
AssertionError: Pagination dropdown label is updated from URL params: expected 'Show 50' to equal 'Show 25'
FAIL: main - Issue-list - Loading URL with stage param loads issues (11008ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".js-filter-button.is-active"}] Unable to locate element: {"method":"css selector","selector":".js-filter-button.is-active"}
FAIL: main - Issue-list - Clicking on a stage filter adds the correct param to the URL (10791ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1)"}
FAIL: main - Issue-list - Toggling a stage filter doesn't leave the param in the URL (10788ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1)"}
FAIL: main - Issue-list - Toggling between stage filters results in last param in URL (10783ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1)"}
FAIL: main - Issue-list - Results are loaded from the query params (10627ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1) a"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1) a"}
FAIL: main - Search (non-auth) - Results are loaded from the query params (10683ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1) a"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1) a"}
FAIL: main - Search (non-auth) - Clicking on label search suggestion works (10987ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-child(1) > div:nth-child(2) > span:nth-child(1) > a:nth-child(1)"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-child(1) > div:nth-child(2) > span:nth-child(1) > a:nth-child(1)"}
FAIL: main - Search (non-auth) - Search from the homepage (16336ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1) a"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1) a"}
FAIL: main - Search (auth) - Results are loaded from the query params (10563ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1) a"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1) a"}
FAIL: main - Search (auth) - Search from the homepage (13971ms)
NoSuchElement: [POST http://localhost:4444/wd/hub/session/e45a5bef-67ff-4170-a05f-50d9df3e0cb9/element / {"using":"css selector","value":".wc-IssueItem:nth-of-type(1) a"}] Unable to locate element: {"method":"css selector","selector":".wc-IssueItem:nth-of-type(1) a"}
```

It could be that you've changed some DOM elements and the tests need updates, but it looks like there's real work needed to get all tests to pass so we don't regress functionality.

hallvors commented:
Yes, for some reason the list of issues doesn't load correctly when the server is running in test mode, although it works fine in non-test mode. Investigating now..
Pushing some fixes addressing your comments so far shortly.

hallvors commented:
As expected, these code changes will change what parameters are sent in the query string, so the hash name of some fixture files will have to change. We'll discuss the details in #834.

hallvors commented:
I'll revert some changes both to take smaller steps and because I need to clarify the distinction between GitHub's search and issues APIs. So just hold the review for now, changes are coming.

miketaylr commented:
OK. It might be best to open up a new PR when you think it's ready for review -- I'm gonna close this one.

Lots of commits plus "undo" commits gets hard to review (and a lot of these commits are missing Issue #s), so I would recommend rebasing into a set of final commits once you think you're in a place that doesn't regress current functionality.

miketaylr commented:
Thinking more about this, I might suggest not doing any more work until we can discuss it in Orlando. This is a significant re-write in order to get label stacking behavior, but we haven't really discussed the relative priority of that feature (or how it should all work, etc.). I don't want to end up wasting your time... so let's talk about what we want to do, and how to split it up into small chunks/issues.
